plugins/router/f5/f5.go

```diff
@@ -0,0 +1,1743 @@
+	}{
+		Member: member,
+	}
+
+	err = f5.post(url, payload, nil)
+	if err != nil {
+		return err
+	}
+
+	members, err := f5.GetPoolMembers(poolname)
+	if err != nil {
+		return err
+	}
+
+	members[member] = true
+
+	glog.V(4).Infof("Added pool member %s to pool %s.",
+		member, poolname)
+
+	return nil
+}
+
+// Delete member from the pool named poolname on F5, and update
+// f5.poolMembers[poolname].
+func (f5 *f5LTM) DeletePoolMember(poolname, member string) error {
+	// The invocation of f5.PoolHasMember has the side effect that it will
+	// initialise f5.poolMembers[poolname], which is used below, if necessary.
+	hasMember, err := f5.PoolHasMember(poolname, member)
+	if err != nil {
+		return err
+	}
+	if !hasMember {
+		glog.V(4).Infof("Pool %s does not have member %s.\n", poolname, member)
+		return nil
+	}
+
+	url := fmt.Sprintf("https://%s/mgmt/tm/ltm/pool/%s/members/%s",
+		f5.host, poolname, member)
+
+	err = f5.delete(url, nil)
+	if err != nil {
+		return err
+	}
+
+	delete(f5.poolMembers[poolname], member)
+
+	glog.V(4).Infof("Pool member %s deleted from pool %s.", member, poolname)
+
+	return nil
+}
+
+// Return f5.routes[policyname], first initializing it from F5 if it is zero.
+func (f5 *f5LTM) getRoutes(policyname string) (map[string]bool, error) {
+	routes, ok := f5.routes[policyname]
+	if ok {
+		return routes, nil
+	}
+
+	url := fmt.Sprintf("https://%s/mgmt/tm/ltm/policy/%s/rules",
+		f5.host, policyname)
+
+	res := f5Policy{}
+
+	err := f5.get(url, &res)
+	if err != nil {
+		return nil, err
+	}
+
+	routes = map[string]bool{}
+
+	for _, rule := range res.Rules {
+		routes[rule.Name] = true
+	}
+
+	f5.routes[policyname] = routes
+
+	return routes, nil
+}
+
+// Check whether the specified route exists for the specified policy.
+func (f5 *f5LTM) routeExists(policyname, routename string) (bool, error) {
+	routes, err := f5.getRoutes(policyname)
+	if err != nil {
+		return false, err
+	}
+
+	return routes[f5UrlEscape(routename)], nil
+}
+
+// Check whether the specified insecure route exists.
+func (f5 *f5LTM) InsecureRouteExists(routename string) (bool, error) {
+	return f5.routeExists(httpPolicyName, routename)
+}
+
+// Check whether the specified secure route exists.
+func (f5 *f5LTM) SecureRouteExists(routename string) (bool, error) {
+	return f5.routeExists(httpsPolicyName, routename)
+}
+
+// Check whether the specified passthrough route exists.
+func (f5 *f5LTM) PassthroughRouteExists(routename string) (bool, error) {
+	routes, err := f5.getPassthroughRoutes()
+	if err != nil {
+		return false, err
+	}
+
+	_, ok := routes[routename]
+
+	return ok, nil
+}
+
+// Add a rule to the specified policy to route requests to the given hostname to
+// the pool named by poolname.
+func (f5 *f5LTM) addRoute(policyname, routename, poolname, hostname,
+	pathname string) error {
+	success := false
+
+	rulesUrl := fmt.Sprintf("https://%s/mgmt/tm/ltm/policy/%s/rules",
+		f5.host, policyname)
+
+	escapedRulename := f5UrlEscape(routename)
```

Comment:
Ignore this - my bad - addRoutes threw me off a bit. See that its adding a rule for a route.